### ISSUE TITLE
Header guard handling conflicts with Google spec

### ISSUE THREAD

**Issue by m3m0ryh0l3** [cpplint/cpplint#44]
According to the [Google guidelines](https://google.github.io/styleguide/cppguide.    html#The__define_Guard), the format of the header guard is:
`{PROJECT}_{FILEPATH}_{FILENAME}_`
e.g. `HELLOWORLD_SRC_INCLUDE_HEADER_H_`

However, CppLint.py insists that the header guard format must be: `{FILEPATH}_{FILENAME}_`

We need a way to specify the $PROJECT

Also, the Google spec specifies that the header guard is an exception to the line-length limits, but cpplint.py does not treat it as an exception.

**Comment by weakcamel** [comment]
Agreed. Although to make it worse, this guide section is inconsistent in its own right:

https://google.github.io/styleguide/cppguide#The__define_Guard

> To guarantee uniqueness, they should be based on the full path in a project's source tree. 
>  For example, the file foo/src/bar/baz.h in project foo should have the following guard:
> `#ifndef FOO_BAR_BAZ_H_`

So according to their own guideline, first sentence, the guard in that example should be one of:

* `FOO_FOO_SRC_BAZ_H_` assumming `foo` is both a project name as well as part of the path, or
* `FOO_SRC_BAR_BAZ_H_` assumming `foo` is a a project name already being *part* of the path

Unless the intention of "based on full path" was not a literal (equals) but loose interpretation, in which case that's not really a specification at all.

**Comment by tkruse** [comment]
Is this already an issue in the https://google.github.io/styleguide/cppguide? If so, please link.

**Comment by weakcamel** [comment]
@tkruse not that I'm aware of.

All I found is https://github.com/google/styleguide/issues/219 which seems to be related (possibly caused by the same inconsistency)

**Comment by GerardMaggiolino** [comment]
This issue is still the first thing that comes up when googling this discrepancy, so for anyone else who finds themselves here a simple solution is to add `--root=..` assuming your project folder is named after your project. 

E.G.
```
cpplint --root=.. --recursive ./src ./include
```
If your project name and top level folder are `foo` and you have `include/baz.hpp` the header guard is now expected to be the more specific `FOO_INCLUDE_BAZ_HPP_`.

**Comment by OliverFrancis** [comment]
In my opinion, the existing format ({FILEPATH}_{FILENAME}_) that is enforced by cpplint.py is better, particularly in projects that make use of git submodules.  For example, let's assume that you have a project hierarchy that resembles the one below:

my_project
|
|___submodule_a
|             file1.h
|             file2.h
|
|___submodule_b
              file3.h

The include guard in file1.h should be SUBMODULE_A_FILE1_H_ because there's no way that the submodule author can know what top-level projects will use the submodule.  Also, given that all header files will be relative to the top-level directory of the project, won't the project name prefix be redundant anyway?

Therefore, I would suggest that google change their style guide to remove the project name prefix and we should leave cpplint.py as it is today.  One minor issue that I have with their format is the lack of an underscore prefix, but that's just my personal taste.

**Comment by Kolin63** [comment]
Correct me if I'm wrong, but I always thought that the purpose of the project name being in the define was to prevent conflicts. For example, if a project had the structure of:
```
my_project/
|
|___submodule_a/
| |___file.h
|
|___submodule_b/
| |___file.h
|
|___file.h

```
then all files would have
```cpp
#ifndef SRC_FILE_H_
#define SRC_FILE_H_

// ...

#endif  // SRC_FILE_H
```
Then, in another file, if you included all three files, you would only be able to access the one you included first. For example, if you did
```cpp
#include "submodule_a/file.h"
#include "submodule_b/file.h"
#include "file.h"
```
you would only get the functions defined in `submodule_a/file.h`.
